Stack/Opc.Ua.Core/Security/Certificates/CertificateTypesProvider.cs

```diff
@@ -0,0 +1,170 @@
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Opc.Ua
+{
+    /// <summary>
+    /// The provider for the X509 application certificates.
+    /// </summary>
+    public class CertificateTypesProvider
+    {
+        /// <summary>
+        /// Create an instance of the certificate provider.
+        /// </summary>
+        public CertificateTypesProvider(ApplicationConfiguration config)
+        {
+            m_securityConfiguration = config.SecurityConfiguration;
+            m_certificateValidator = new CertificateValidator();
+            m_certificateValidator.Update(m_securityConfiguration).GetAwaiter().GetResult();
+            // for application certificates, allow untrusted and revocation status unknown, cache the known certs
+            m_certificateValidator.RejectUnknownRevocationStatus = false;
+            m_certificateValidator.AutoAcceptUntrustedCertificates = true;
+            m_certificateValidator.UseValidatedCertificates = true;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the application should send the complete certificate chain.
+        /// </summary>
+        /// <remarks>
+        /// If set to true the complete certificate chain will be sent for CA signed certificates.
+        /// </remarks>
+        public bool SendCertificateChain => m_securityConfiguration.SendCertificateChain;
+
+        /// <summary>
+        /// Return the instance certificate for a security policy.
+        /// </summary>
+        /// <param name="securityPolicyUri">The security policy Uri</param>
+        public X509Certificate2 GetInstanceCertificate(string securityPolicyUri)
+        {
+            if (securityPolicyUri == SecurityPolicies.None)
+            {
+                // return the default certificate for None
+                return m_securityConfiguration.ApplicationCertificates.FirstOrDefault().Certificate;
+            }
+            var certificateTypes = CertificateIdentifier.MapSecurityPolicyToCertificateTypes(securityPolicyUri);
+            foreach (var certType in certificateTypes)
+            {
+                var instanceCertificate = m_securityConfiguration.ApplicationCertificates.FirstOrDefault(id => id.CertificateType == certType);
+                if (instanceCertificate == null &&
+                    certType == ObjectTypeIds.RsaSha256ApplicationCertificateType)
+                {
+                    instanceCertificate = m_securityConfiguration.ApplicationCertificates.FirstOrDefault(id => id.CertificateType == null);
+                }
+                if (instanceCertificate == null &&
+                    certType == ObjectTypeIds.ApplicationCertificateType)
+                {
+                    instanceCertificate = m_securityConfiguration.ApplicationCertificates.FirstOrDefault();
+                }
```

Comment:
need to add this line here to return default https cert:
if (instanceCertificate == null &&
    certType == ObjectTypeIds.HttpsCertificateType)
{
    instanceCertificate = m_securityConfiguration.ApplicationCertificates.FirstOrDefault();
}